File: stack/stack.exs
defmodule Stack do

    defstruct items: []

    def new do
        {:ok, %Stack{}}
    end

    def push(%Stack{items: items}, item) do
        {:ok, %Stack{ items: [item] ++ items }}
    end

    def pop(%Stack{items: []}) do
        raise("Cannot pop an empty Stack")
    end

    def pop(%Stack{items: [head | tail]}) do
        {:ok, head, %Stack{items: tail}}
    end

    def top(%Stack{items: []}) do 
        raise("Cannot see the top of an empty Stack")
    end

    def top(%Stack{items: [head | _tail]}) do
        {:ok, head}
    end

end

File: linked-stack/python/Stack.py
from Item import Item

class Stack:

    def __init__(self, max: int = 0):
        self.max = max
        self.itemCnt = 0
        self.item = None

    def full(self):
        return self.itemCnt == self.max

    def pop(self):
        if isinstance(self.item, Item):
            tmpItem = self.item

            self.itemCnt -= 1
            self.item = tmpItem.link

            return tmpItem.value
        else:
            raise(BaseException("Cannot pop an empty stack"))

    def push(self, value):
        if self.full():
            raise(BaseException("Cannot push into a full stack"))

        tmpItem = Item(value, self.item)

        self.itemCnt += 1
        self.item = tmpItem

        return self

    def top(self):
        if not isinstance(self.item, Item):
            raise(BaseException("Cannot top an empty stack"))
        
        return self.item.value
